Basic jemalloc command for printing arenas info with bin

jetchirag opened the pull request:
<!-- Please make sure to read the testing and linting instructions at https://github.com/pwndbg/pwndbg/blob/dev/DEVELOPING.md before creating a PR -->
For #2174

First command for jemalloc to print basic arenas details, it is not modularized and currently only for basic testing purpose.

```
pwndbg> arenas_info
Number of arenas:  1
Arena Address:  0x7ffff78010c0
Index   Address Slabcur
0 0x7ffff7814528 0x7ffff7816980
1 0x7ffff7814608 0x0
2 0x7ffff78146e8 0x0
...
```

jetchirag committed:
Basic jemalloc arenas info command

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
linter

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Hex address

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

CptGibbon commented:
Awesome!
I'll get jemalloc installed and take a look at this 👍

CptGibbon commented:
@jetchirag 
Huh I'm seeing the below error when I try the new command whilst debugging the example program from #2174 
What am I doing wrong?

<img width="729" alt="Screenshot 2024-05-24 at 09 47 34" src="https://github.com/pwndbg/pwndbg/assets/16000770/587c9fcc-f914-48d6-bf6d-79d57cf2a7cd">

My jemalloc version is 5.3.0-182-gf9c0b5f7f8a917661db39289e38ec94d9d198f11

jetchirag commented:
@CptGibbon Are you on dev branch? It was renamed, I'm using [5.3.0](https://github.com/jemalloc/jemalloc/releases/tag/5.3.0). Btw to update on progress, I'm currently working on parsing rtree which stores extent metadata, will add a command for that soon.

CptGibbon commented:
> Are you on dev branch?

I'm on the `jemalloc_1` branch that this PR is based on.

EDIT: Oh I guess you meant dev branch of jemalloc?
In which case: yes.
Which branch should I use? I see `master`, `stable-3` & `stable-4`, but they're all considered "stale" with their last update at least a couple of years ago.

jetchirag commented:
> > Are you on dev branch?
> 
> I'm on the `jemalloc_1` branch that this PR is based on.
> 
> EDIT: Oh I guess you meant dev branch of jemalloc? In which case: yes. Which branch should I use? I see `master`, `stable-3` & `stable-4`, but they're all considered "stale" with their last update at least a couple of years ago.

The latest release, version 5.3.0.

Edit: you can download it from https://github.com/jemalloc/jemalloc/releases/tag/5.3.0

CptGibbon commented:
Ah okay thanks 👍
<img width="247" alt="Screenshot 2024-05-24 at 14 25 52" src="https://github.com/pwndbg/pwndbg/assets/16000770/9c4834f0-84a9-4d40-a3ee-1ae158591192">

CptGibbon commented:
Great stuff, just a couple of notes for this PR:

1) I think renaming the command to something with "jemalloc" in the title makes it clear which allocator it's for (this will be cumbersome for now but if we implement automatic allocator detection later it could change to something more terse)
2) Specify somewhere which version of jemalloc this is for (perhaps just print it in the command output for now, we can do version detection later) that way people won't be confused like I was if it doesn't work
3) Amend `setup-dev.sh` so that reviewers don't have to manually install jemalloc and build your test program

jetchirag committed:
Added jemalloc to dev setup script and WIP code foor rtree parsing

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Lint changes

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Lint changes 2

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag commented:
@CptGibbon Added prefixes to command, and also installation script. Also pushed some WIP code for rtree parsing.

CptGibbon commented:
> @CptGibbon Added prefixes to command, and also installation script. Also pushed some WIP code for rtree parsing.

Okay thanks, great work 👍
I would recommend creating tests for these new jemalloc commands before adding any more to this PR.
Let me know if you want help writing them.

CptGibbon commented:
I'm seeing a few lint commits, I find working in an IDE that supports devcontainers can alleviate this.
pwndbg has a devcontainer setup that will automatically install the appropriate formatting & linting extensions and run them each time you save a file.

jetchirag committed:
Removed old commands, added jemalloc related class to generate bins and extent data

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag commented:
So I have removed the old commands and added proper class for generating memory mapping by jemalloc. There are lots of todos, I'll be adding one or two classes for rtree (including one to finding extent information from memory pointer address given by `malloc()`) and other structures then move onto actual command followed by tests as I'd also like to get started with test early since there's that discussed problem of testing memory addresses.

Things were slow past 2 weeks, I did read a lot of documentations on the same and will try to pick up pace now.

Also, thank you for dev containers suggestion though seems like it didn't fix linting on `setup-dev.sh` or maybe I forgot to run linting on it. It's just whitespace, I'll fix that in next commit, hopefully

jetchirag committed:
Added RTree class

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Uncommented print lines for testing

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Added mask function from jemalloc

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Added heap command

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag commented:
@CptGibbon I think now we have something which resembles `malloc_chunk` and `heap` command which is the primary goal. The structure's still not there yet and tests aren't done but it would be great if I can get feedback on this so we can make changes early.

CptGibbon commented:
> @CptGibbon ... it would be great if I can get feedback on this so we can make changes early.

No problem, I'll review today.
Thanks for the update 🙏

CptGibbon commented:
@jetchirag
I loaded your test program into GDB and ran your new commands.
Screens of my experience below:

### jemalloc_heap
<img width="313" alt="Screenshot 2024-06-20 at 15 56 46" src="https://github.com/pwndbg/pwndbg/assets/16000770/ec4b0a5c-da19-4de7-97e2-4fb7e56da4f0">

`jemalloc_heap` appears to work, though some of those addresses look spurious, can you confirm this is correct behavior?

### find_extent
<img width="335" alt="Screenshot 2024-06-20 at 16 02 27" src="https://github.com/pwndbg/pwndbg/assets/16000770/9fce8e73-37d2-46ea-b961-b7256bf804c2">

`find_extent` appears to work, again could you confirm this is correct behavior?

### extent_info
Used by the `jemalloc_heap` and `find_extent` commands.

### jemalloc_base_info
<img width="735" alt="Screenshot 2024-06-20 at 16 06 04" src="https://github.com/pwndbg/pwndbg/assets/16000770/2d43b7a9-f9ae-4a47-beb9-0ff12954785c">

No dice.

### get_extent
<img width="735" alt="Screenshot 2024-06-20 at 16 08 48" src="https://github.com/pwndbg/pwndbg/assets/16000770/bd2543d7-859f-4c86-b094-d7be60a9e7fc">

Not sure if I'm using the correct address as an argument here but nothing I try works...

### jemalloc_test
<img width="735" alt="Screenshot 2024-06-20 at 16 09 45" src="https://github.com/pwndbg/pwndbg/assets/16000770/60696288-c118-49b1-8860-683e8a42397e">

One last note; if you want your test program to get linked with jemalloc you'll need to add a corresponding line to the makefile that builds tests.

jetchirag commented:
Thank you for testing @CptGibbon 
Currently only `jemalloc_heap`,`jemalloc_find_extent`and `jemalloc_extent_info` are to be used, rest are just for testing purpose. Since I've changed the structure a lot, some old test ones would have broke, it's fine.

Your output for heap and find extent seems to be expected. Yes there are some false addresses, some of it seems to be due to cache_oblivious, I'll investigate that and add note in output if needed or fix it. Any comment on the code itself?

CptGibbon commented:
> Any comment on the code itself?

I'm usually of the mind that if it works it works, but I can take a closer look 👍

So far my only concern aside from the address issue is that `jemalloc_heap` was rather slow to complete, at around 3 seconds. This was running on a free codespaces instance of the smallest variety, but that's still rather slow compared to most commands.

Great work 👌

jetchirag commented:
> So far my only concern aside from the address issue is that jemalloc_heap was rather slow to complete, at around 3 seconds. This was running on a free codespaces instance of the smallest variety, but that's still rather slow compared to most commands.

Yes that has been my concern as well. I've mentioned this in comment, it's happening due to the size of rtree. I was trying to find it there's an alternate way, perhaps it maintains caches or thread specific information, will check.

CptGibbon commented:
Just 1 note and 1 question:

A current development goal is to uncouple as much code from the gdb library as possible, hopefully this makes testing easier and makes room for different debugger backends.

One step towards this is `gdblib.memory.fetch_struct_as_dictionary()`
https://github.com/pwndbg/pwndbg/blob/f21000356f819f5074e9a10ed74dedda89243da5/pwndbg/gdblib/memory.py#L388-L397

Which will eventually become something like `debugger.fetch_struct_as_dictionary()`. 

An example use can be found in `gdblib.heap.ptmalloc.fetch_chunk_metadata()`.

The transition to debugger-agnostic code is still underway, so if using the above function would slow you down too much it's not a show-stopper, but it's something to consider.

I see you credit jegdb in the source, which code specifically is taken from jegdb?

jetchirag commented:
@CptGibbon I'll check that function. I don't think there's much code taken directly from that script. The code for bit shifting bitfield was directly used 

https://github.com/pwndbg/pwndbg/pull/2176/files#diff-c96fafbf3d980a07d7f741d99185b64e1a09dd7371ac617dc6f7bffb25b7f68bR380-R389

and `rtree_leaf_elm_bits_extent_get`

Primarily I used the script as base to correlate with jemalloc source code to understand structure.

jetchirag commented:
> One last note; if you want your test program to get linked with jemalloc you'll need to add a corresponding line to the makefile that builds tests.

Is it fine to directly link jemalloc path as such in makefile:
```
heap_jemalloc_extent_info.out: heap_jemalloc_extent_info.c
	@echo "[+] Building heap_jemalloc_extent_info.out"
	${CC} -g -O0 -o heap_jemalloc_extent_info.out heap_jemalloc_extent_info.c -pthread -lpthread /usr/local/lib/libjemalloc.a -lm -lstdc++ -pthread
```

jetchirag committed:
First jemalloc test - experiment

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

CptGibbon commented:
> Is it fine to directly link jemalloc path as such in makefile:

Yes that seems to be the current project standard.

@disconnect3d are you happy with this approach?

jetchirag commented:
I noticed the duplicate flag in makefile, will remove that

jetchirag commented:
I looked at cache stores in thread specific information structure and elsewhere but we won't be able to use it for finding extents. I think I will have to stick to current method for now, unless someone has a better approach.

I'll utilize the function you linked to. I'll have to compile jemalloc with cache oblivious disabled to check those weird addresses in heap command.

jetchirag commented:
Can you please also answer if these commands (just the functionality) fulfil the primary scope of this project?

CptGibbon commented:
> I'll utilize the function you linked to.

Great!

> I'll have to compile jemalloc with cache oblivious disabled to check those weird addresses in heap command.

Okay thanks, let us know if that changes anything.

> Can you please also answer if these commands (just the functionality) fulfil the primary scope of this project?

I think these fulfill the jemalloc portion of the project 👍
Once we've ironed out the existing issues we can ask the maintainers to merge it and you're welcome to move onto the kernel portion of the project.

Thanks for your hard work so far 🙏

jetchirag committed:
some refactoring and tests

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
utilizing debugger-agnostic functions

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
merge upstream

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
use new api where possible

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Test run pre mypy fixes

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
minor fix for makefile

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Use regex match for address in test find extent

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
minor test change with lowered allocated memory

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Don't hardcode address for ptr in test

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Fix ptr not in breakpoint in test

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Cache oblivious adds 4KiB to allocations

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Corrected size in test

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Remove duplicate extents while parsing rtree

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Skip heap test

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Merge remote-tracking branch 'upstream/dev' into jemalloc_1

jetchirag commented:
@CptGibbon mypy is not sure about the type of sub dicts in GdbDict from that struct as dictionary function and thus there are several error messages in linting. Is there any preferred way to deal with this?

```
pwndbg/gdblib/heap/jemalloc.py:383: error: Item "dict[str, GdbDict | int]" of "GdbDict | int" has no attribute "address"  [union-attr]
                    bins_addr = int(self._Value["bins"].address)
```

CptGibbon commented:
> @CptGibbon mypy is not sure about the type of sub dicts in GdbDict from that struct as dictionary function and thus there are several error messages in linting

That sub dict is a @gsingh93 creation so we'd best ask them.

disconnect3d commented on .devcontainer/devcontainer.json:
```diff
@@ -55,7 +55,8 @@
     "runArgs": [
-        "--cap-add=SYS_PTRACE"
+        "--cap-add=SYS_PTRACE",
+        "--privileged"
```
Why do we need this?

jetchirag replied:
This was added to disable ASLR which was failing for me locally. I'll remove/git ignore it.

disconnect3d replied:
Please remove this flag. One can disable ASLR globally outside of the container if needed
```suggestion
```

disconnect3d commented on pwndbg/commands/heap.py:
```diff
@@ -1546,3 +1547,74 @@ def heap_config(filter_pattern: str) -> None:
+@pwndbg.commands.ArgparsedCommand(parser, category=CommandCategory.HEAP)
+def jemalloc_find_extent(addr) -> None:
+    print(C.banner("Jemalloc find extent"))
+    print("This command only support jemalloc 5.3.0")
```
5.3.0 or newer too? or we dont know for sure?

Maybe: "(This command was tested only for jemalloc 5.3.0)"?

jetchirag replied:
What about "This command was tested only for jemalloc 5.3.0 and does not support lower versions"? Since we know for sure lower versions are not supported due to different structure.

disconnect3d commented on pwndbg/commands/heap.py:
```diff
@@ -1546,3 +1547,74 @@ def heap_config(filter_pattern: str) -> None:
+# Jemalloc
+
+parser = argparse.ArgumentParser(
+    description="Returns extent information for pointer address allocated by jemalloc"
```
> Memory is conceptually broken into extents. Extents are always aligned to multiples of the page size. This alignment makes it possible to find metadata for user objects quickly. User objects are broken into two categories according to size: small and large. Contiguous small objects comprise a slab, which resides within a single extent, whereas large objects each have their own extents backing them.

hmm... wonder if we can at least add a comment on those extents somewhere (aka what are they lol) or inform the user about them, idk...

jetchirag replied:
Added information on extent in comments

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,488 @@
+LG_VADDR = 48
+LG_PAGE = 12
+RTREE_NLIB = LG_PAGE
+# jemalloc/include/jemalloc/internal/jemalloc_internal_types.h
```
Please change to a link to github (or other VCS) with specific version this is based from (5.3.0), or even commit

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,488 @@
+RTREE_NHIB = (1 << (LG_SIZEOF_PTR + 3)) - LG_VADDR
+
+
+# TODO: Move to relevant place
```
What is relevant place?

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,488 @@
+    return ((1 << current_field_width) - 1) << current_field_shift
+
+
+LG_QUANTUM = 4  # TODO: lookup value acc to architecture from include/jemalloc/internal/quantum.h (currently set for arch64)
```
The arch name is 'aarch64' or 'arm64'. But anyway, can we make this work for x86-64 as well?

jetchirag replied:
According to https://github.com/jemalloc/jemalloc/blob/5.3.0/include/jemalloc/internal/quantum.h#L21, value for it is 4 as well so we can remove this TODO.

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,488 @@
+SC_LG_TINY_MIN = 3
+SC_NTINY = LG_QUANTUM - SC_LG_TINY_MIN
+
+SC_LG_NGROUP = 2
```
Do we know what does SC stand for? Can we add a comment for it?

jetchirag replied:
Added

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,488 @@
+
+    @property
+    def root(self):
+        # return self.rtree["root"]
```
?

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,488 @@
+        node = pwndbg.gdblib.memory.fetch_struct_as_dictionary("rtree_node_elm_s", addr)
+
+        if int(node["child"]["repr"]) == 0:
+            return None
```
What does this mean? Shall we comment it?

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,488 @@
+        # leaf = pwndbg.gdblib.memory.poi(rtree_leaf_elm_s, addr)
+        leaf = pwndbg.gdblib.memory.fetch_struct_as_dictionary("rtree_leaf_elm_s", addr)
+        if leaf["le_bits"]["repr"] == 0:
+            return None
```
What does this mean? Shall we comment it?

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,488 @@
+                    )
+                    node = pwndbg.gdblib.memory.pack_struct_into_dictionary(fetched_struct)
+
+                    if node["child"]["repr"] == 0:
```
What does this mean? Shall we comment it?

disconnect3d commented:
Hi, I added a few comments here and there. It would be nice to document as much "why" and "what" as possible for further readers or maintainers of the code, as otherwise, it may be extremely hard to take this work upon and extend it or adjust for other jemalloc versions or needs etc.

Ideally, this could also explain some implementation details of jemalloc - aka how it allocates memory, what are extents, design choices etc etc.

jetchirag committed:
More explanation

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Added State name to extent info

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Added more comments

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag committed:
Details about the lookup function

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag commented:
@disconnect3d @CptGibbon I've added more comments describing most of the code and jemalloc itself.

CptGibbon commented:
Thanks for the latest update and comments 🙏

Here's my experience using the new commands:

### jemalloc_heap

![jemalloc_heap](https://github.com/user-attachments/assets/150301ec-7a01-43a9-b6f9-c236e52327a5)

How come each extent is printed twice?
I'm not seeing any spurious addresses any more though 👍
It would be great to have a "small class: True" example for the sake of tests.

### jemalloc_extent_info

![jemalloc_extent_info](https://github.com/user-attachments/assets/0d590053-c9aa-487c-927c-22b8aaeb9d82)

Looks good 👍
Any reason the version warning is different from the other two commands?

### jemalloc_find_extent

![jemalloc_find_extent](https://github.com/user-attachments/assets/c07f0dd4-107c-4c00-a41c-adbd3860f3f7)

Am I using this incorrectly? How come the extent address is off by 2? That didn't happen before...

I think this work is so nearly done, just a couple more things to square away.

jetchirag commented:
@CptGibbon Different message was unintentional.

For the last issue, the allocated address is incorrect so there is some issue, it couldn't locate correct extent address. I'm checking why that happens.

jetchirag commented:
@CptGibbon Can you please share the code file you ran to get these outputs? How much memory did you allocate?

CptGibbon commented:
> @CptGibbon Can you please share the code file you ran to get these outputs? How much memory did you allocate?

I used your test files, I believe the output above is from `tests/gdb-tests/tests/binaries/heap_jemalloc_heap.c`

jetchirag committed:
Fixed double printing issue and some extents missing in heap

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag commented:
> How come each extent is printed twice?

Accidentally had it printing it twice in some cases. Fixed.


> It would be great to have a "small class: True" example for the sake of tests

Yes, it was there but missing in heap due to last changes. Fixed and now visible.

jetchirag committed:
Add check for addr alignment in lookup hard

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

jetchirag commented:
Sample outputs:

```
pwndbg> jemalloc_heap
Jemalloc heap
This command was tested only for jemalloc 5.3.0 and does not support lower versions

Allocated Address: 0x7ffff7000000
Extent Address: 0x7ffff7a16a00
Size: 0x400000
Small class: False
State: Active

Allocated Address: 0x7ffff7800000
Extent Address: 0x7ffff7a16500
Size: 0x7000
Small class: False
State: Active

Allocated Address: 0x7ffff7807000
Extent Address: 0x7ffff7a16580
Size: 0x1000
Small class: True
State: Active

Allocated Address: 0x7ffff7808000
Extent Address: 0x7ffff7a16980
Size: 0x1f8000
Small class: False
State: Retained
```

```
pwndbg> jemalloc_find_extent ptr2
Jemalloc find extent
This command was tested only for jemalloc 5.3.0 and does not support lower versions

Pointer Address: 0x7ffff7000000
Extent Address: 0x7ffff7a16a00

Allocated Address: 0x7ffff7000000
Extent Address: 0x7ffff7a16a00
Size: 0x400000
Small class: False
State: Active

pwndbg> jemalloc_find_extent ptr
Jemalloc find extent
This command was tested only for jemalloc 5.3.0 and does not support lower versions

Pointer Address: 0x7ffff7807000
Extent Address: 0x7ffff7a16580

Allocated Address: 0x7ffff7807000
Extent Address: 0x7ffff7a16580
Size: 0x1000
Small class: True
State: Active
```

```
pwndbg> jemalloc_extent_info 0x7ffff7a16580
Jemalloc extent info
This command was tested only for jemalloc 5.3.0 and does not support lower versions

Allocated Address: 0x7ffff7807000
Extent Address: 0x7ffff7a16580
Size: 0x1000
Small class: True
State: Active
```

jetchirag committed:
Removed prints for testing

Signed-off-by: Chirag Aggarwal <thechiragaggarwal@gmail.com>

CptGibbon commented:
@jetchirag thank you for the latest update

If you're happy with the state of this work, you can take it out of draft state and we'll give it a final review 👍

jetchirag commented:
@CptGibbon Moved out of draft. There might be changes needed on structuring the code since everything is currently in same file. Let me know if any.

CptGibbon reviewed (approved):
I think this satisfies the GSoC criteria set out earlier this year.
It gives pwndbg a foot in the door of jemalloc inspection and I'm sure this feature will see plenty of use. I look forward to seeing future contributions to it by yourself & others.

There have been a lot of changes to how pwndbg fundamentally operates during this working period (my chance to encourage more continuous integration style contributions) so I defer to the honorable @disconnect3d for any pwndbg-specific feedback.

Thank you @jetchirag for your hard work and for continuing despite losing your laptop!

jetchirag commented:
@CptGibbon Thank you, I'll remove merge conflicts in next commit.

disconnect3d committed:
Update .devcontainer/devcontainer.json

disconnect3d commented on pwndbg/gdblib/heap/jemalloc.py:
```diff
@@ -0,0 +1,555 @@
+
+                        self._extents.append(extent_tmp)
+
+            except gdb.MemoryError:
```
@jetchirag is there any way we can do  this elsewhere? I mean, we are excepting a huge block of code here :/

disconnect3d commented:
Btw the commands are a bit slow, is there a way to cache it or make it faster? Why is it so slow?

CptGibbon commented:
> Btw the commands are a bit slow, is there a way to cache it or make it faster? Why is it so slow?

I believe parsing the rtree is rather costly.
My two cents is that this feature works and performance wasn't part of the original spec, so perhaps we could let the author continue working on their kernel feature which is due shortly and then revisit this feature to discuss improvements.

disconnect3d committed:
Move jemalloc.py aglib/heap/jemalloc.py

disconnect3d committed:
Merge branch 'dev' into jemalloc_1

disconnect3d commented:
Commited a merge commit here to (hopefully) fix conflicts.

disconnect3d commented:
@jetchirag Ping, do u maybe know why the tests fail on CI?

disconnect3d committed:
Update test_heap.py: fix jemalloc support string

disconnect3d commented:
Ok, the tests were failing bcoz this string was updated in code, but not in tests:

Old:
> This command only support jemalloc 5.3.0

New:
> This command was tested only for jemalloc 5.3.0 and does not support lower versions

I changed this in the tests now.

disconnect3d commented on pwndbg/aglib/heap/jemalloc.py:
```diff
@@ -0,0 +1,555 @@
+        self._slabs = None
+
+    @property
+    def slabs(self):
```
What is this property for? It seems to be unused across the code and it seems to be broken: it uses `self._bins` which is not defined in the `Arena` class.

cc: @jetchirag

jetchirag replied:
@disconnect3d I left this for future extension as I started working on this project with this property but it's not needed right now. I'll remove Arenas.

disconnect3d committed:
Update jemalloc.py: fix mypy issues

disconnect3d commented:
Fixed mypy issues in jemalloc.py. Also, it seems that its `Arena` class along its `slabs` property are unused. Is this intended? Do we plan to extend it in the future or what? Shall we remove it?

cc: @jetchirag @CptGibbon
